src/utils/skip_list.hpp

```diff
@@ -1495,6 +1642,146 @@ class SkipList final : detail::SkipListNode_base {
     }
   }
 
+  /// Creates chunks for parallel processing of the skip list.
+  /// Uses the maximum layer for efficient chunking, splitting based on max-height elements.
+  /// Ensures complete coverage by making first chunk start from beginning and last chunk end at end.
+  /// This method is thread-safe and can be called concurrently.
+  ///
+  /// @param num_chunks The number of chunks to create
+  /// @return ChunkCollection containing the chunks
+  ChunkCollection create_chunks(size_t num_chunks) const {
+    if (head_ == nullptr) {
+      // Return one empty chunk for empty list
+      return {{Chunk{nullptr, nullptr}}};
+    }
+
+    std::array<TNode *, kSkipListMaxHeight> start{};
+    std::array<TNode *, kSkipListMaxHeight> end{};
+    for (int layer = 0; layer < kSkipListMaxHeight; ++layer) {
+      start[layer] = head_->nexts[layer].load(std::memory_order_acquire);
+      end[layer] = nullptr;
+    }
+    return create_chunks_(num_chunks, start, end);
+  }
+
+  template <typename TKey>
+  ChunkCollection create_chunks(size_t num_chunks, const std::optional<TKey> &lower_bound,
+                                const std::optional<TKey> &upper_bound) const {
+    if (head_ == nullptr) {
+      // Return one empty chunk for empty list
+      return {{Chunk{nullptr, nullptr}}};
+    }
+
+    if (lower_bound && upper_bound && lower_bound.value() > upper_bound.value()) {
+      // Invalid range
+      return {{Chunk{nullptr, nullptr}}};
+    }
+
+    std::array<TNode *, kSkipListMaxHeight> preds{};
+    std::array<TNode *, kSkipListMaxHeight> succs{};
+    std::array<TNode *, kSkipListMaxHeight> start{};
+    std::array<TNode *, kSkipListMaxHeight> end{};
+
+    if (lower_bound) {
+      auto found_layer = find_node<GCPolicy::DoNotRun>(lower_bound.value(), preds, start);
+      // Lower bound not found
+      if (found_layer == -1 && start[0] == nullptr) {
+        return {{Chunk{nullptr, nullptr}}};
+      }
+    } else {
+      for (int layer = 0; layer < kSkipListMaxHeight; ++layer) {
+        start[layer] = head_->nexts[layer].load(std::memory_order_acquire);
+      }
+    }
+
+    if (upper_bound) {
+      find_node<GCPolicy::DoNotRun>(upper_bound.value(), end, succs);
+      for (int layer = 0; layer < kSkipListMaxHeight; ++layer) {
+        // head_ is not a valid node (use next)
+        auto *current = end[layer] == head_ ? head_->nexts[layer].load(std::memory_order_acquire) : end[layer];
+        // Find the first element over the bound
+        while (current != nullptr && (current->obj < upper_bound.value() || current->obj == upper_bound.value())) {
+          if (!current->marked.load(std::memory_order_acquire)) {
+            while (!current->fully_linked.load(std::memory_order_acquire))
```

Comment:
wouldn't it be better to use wait(false) on fully_linked and add notify in the code instead of busy loops